File: car/scripts/csv_to_json.js
#!/usr/bin/env node
/**
 * Convert registered_vehicle_maintenance_workshops.csv to JSON for fast loading.
 * Run: node car/scripts/csv_to_json.js
 */

const fs = require("fs");
const path = require("path");

const projectRoot = path.resolve(__dirname, "../..");
const csvPath = path.join(projectRoot, "registered_vehicle_maintenance_workshops.csv");
const jsonPath = path.join(projectRoot, "car", "workshops.json");

function parseCSV(text) {
  const rows = [];
  let i = 0;
  while (i < text.length) {
    const row = [];
    while (i < text.length) {
      let cell = "";
      if (text[i] === '"') {
        i++;
        while (i < text.length) {
          if (text[i] === '"') {
            i++;
            if (text[i] === '"') {
              cell += '"';
              i++;
            } else break;
            }
          } else {
            cell += text[i++];
          }
        }
      } else {
        while (i < text.length && text[i] !== "," && text[i] !== "\n" && text[i] !== "\r") {
          cell += text[i++];
        }
      }
      row.push(cell);
      if (text[i] === ",") {
        i++;
        continue;
      }
      if (text[i] === "\n" || text[i] === "\r") {
        if (text[i] === "\r" && text[i + 1] === "\n") i++;
        i++;
        break;
      }
      if (i >= text.length) break;
    }
    if (row.length > 0 && row.some((c) => c !== "")) rows.push(row);
  }
  return rows;
}

const csvText = fs.readFileSync(csvPath, "utf8");
const lines = parseCSV(csvText);
const headers = lines[0];
const rows = lines.slice(1).map((cells) => {
  const obj = {};
  headers.forEach((h, i) => {
    obj[h] = cells[i] ?? "";
  });
  return obj;
});

fs.writeFileSync(jsonPath, JSON.stringify(rows), "utf8");
console.log(`Wrote ${rows.length} workshops to ${jsonPath}`);

File: car/scripts/csv_to_json.py
#!/usr/bin/env python3
"""Convert registered_vehicle_maintenance_workshops.csv to JSON for fast loading."""

import csv
import json
from pathlib import Path

CSV_PATH = Path(__file__).resolve().parent.parent.parent / "registered_vehicle_maintenance_workshops.csv"
JSON_PATH = Path(__file__).resolve().parent.parent / "workshops.json"


def main():
    rows = []
    with open(CSV_PATH, newline="", encoding="utf-8") as f:
        reader = csv.DictReader(f)
        for row in reader:
            rows.append(row)
    with open(JSON_PATH, "w", encoding="utf-8") as f:
        json.dump(rows, f, ensure_ascii=False)
    print(f"Wrote {len(rows)} workshops to {JSON_PATH}")


if __name__ == "__main__":
    main()
